### ISSUE TITLE
Question about LightAttention input shape in TransferModel.forward

### ISSUE THREAD

**Issue by DRanger666** [Kuhlman-Lab/ThermoMPNN#32]
I had a question about the LightAttention implementation versus my reading of the paper description.

The paper describes the LA block as using self-attention to reweight the residue-environment vector, e.g. “reweight the input vector using contextual information via self-attention” and “self-attention to reweight the vector based on learned context.”

In `transfer_model.py`, the main `TransferModel.forward` path appears to build one per-mutation vector and then pass it to `LightAttention` after adding singleton batch/sequence axes:

```python
lin_input = torch.cat(inputs, -1)
lin_input = torch.unsqueeze(torch.unsqueeze(lin_input, -1), 0)
lin_input = self.light_attention(lin_input, mask)
```

Inside `LightAttention`, softmax is applied over the last axis:

```python
self.softmax = nn.Softmax(dim=-1)
o1 = o * self.softmax(attention)
```

My understanding is that, in this path, `lin_input` has shape approximately `[1, 384, 1]`, so `softmax(..., dim=-1)` is over a singleton axis and returns all ones. A small local probe with the released `2OCJ` example and `thermoMPNN_default.pt` gave:

```text
light_input_shape=(1, 384, 1)
attention_weights_min=1.00000000
attention_weights_max=1.00000000
input_output_allclose=False
feature_conv_output_allclose=True
```

So the LA block is not an identity map: the convolutional path changes the vector. However, in this `TransferModel.forward` path, the softmax-normalized attention term appears shape-constrained to a constant weighting.

Am I missing another intended path where `LightAttention` receives a non-singleton sequence axis, or is this the intended implementation? If intended, should the LA block be interpreted here mainly as a learned convolutional transformation of the 384-vector rather than attention-based gating?

**Comment by hdieckhaus** [comment]
Hi @DRanger666 ,

You are correct, this is a mistake that I made in the original ThermoMPNN code. I didn’t notice it until after the model and paper was already published, but the LightAttention in ThermoMPNN is not implemented correctly.

It was my intention to apply it across the embedding dimension (384), but I accidentally applied it to the last dimension, which you noted. So, the LightAttention block becomes pretty much just a convolution. 

However, there is one weird consequence of this error: the kernel size on the LA convolutions is 9, not 1, so every convolution gets 4 padded zeroes on either side, effectively adding a ton of noise (and a lot of parameters). Perhaps this whole mess is why the LA block doesn’t actually help very much (like 1-2%), but weirdly, I think the noise helped it avoid overfitting like I saw when unfreezing more of the layers. In any case, this is not how the original LightAttention paper intended for this to work, so it is a misnomer to call it that.

In the later ThermoMPNN-D, I addressed this by reducing the LA kernel size to 1, meaning it actually behaves like a set of normal linear layers, and I got very similar performance with far fewer parameters, since I wasn’t wasting a bunch on the zeroed padding. I think you could do the same thing with a bigger MLP or a proper attention layer, rather than this weird LightAttention setup. 

Hope this helps clear things up!
